src/lib/support/Pool.h

```diff
@@ -0,0 +1,139 @@
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+/**
+ *  @brief
+ *   A class template used for allocating Objects.
+ *
+ *  @tparam     T   a subclass of Object to be allocated.
+ *  @tparam     N   a positive integer max number of objects the pool provides.
+ */
+template <class T, size_t N>
+class BitMapObjectPool
+{
+public:
+    BitMapObjectPool()
+    {
+        for (size_t word = 0; word * kBitChunkSize < N; ++word)
+        {
+            mUsage[word].store(0);
+        }
+    }
+
+    static size_t Size() { return N; }
+
+    template <typename... Args>
+    T * New(Args &&... args)
+    {
+        for (size_t word = 0; word * kBitChunkSize < N; ++word)
+        {
+            auto & usage = mUsage[word];
+            auto value   = usage.load();
+            for (size_t offset = 0; offset < kBitChunkSize && offset + word * kBitChunkSize < N; ++offset)
+            {
+                if ((value & (kBit1 << offset)) == 0)
+                {
+                    if (usage.compare_exchange_strong(value, value | (kBit1 << offset)))
+                    {
+                        return new (GetPoolHead() + (word * kBitChunkSize + offset)) T(std::forward<Args>(args)...);
+                    }
+                    else
+                    {
+                        value = usage.load(); // if there is a race, update new usage
+                    }
+                }
+            }
+        }
+        return nullptr;
+    }
+
+    void Delete(T * obj)
+    {
+        if (obj == nullptr)
+            return;
+
+        size_t at     = static_cast<size_t>(obj - GetPoolHead());
+        size_t word   = at / kBitChunkSize;
+        size_t offset = at - (word * kBitChunkSize);
+
+        // ensure the obj is in the pool
+        assert(at >= 0);
+        assert(at < N);
+
+        obj->~T();
+
+        auto value = mUsage[word].fetch_and(~(kBit1 << offset));
+        assert((value & (kBit1 << offset)) != 0); // assert fail when free an unused slot
+    }
+
+#if !defined(NDEBUG)
+    template <typename F>
+    void ForEachActiveObject(F f) const
+    {
+        for (size_t word = 0; word * kBitChunkSize < N; ++word)
+        {
+            auto & usage = mUsage[word];
+            auto value   = usage.load();
+            for (size_t offset = 0; offset < kBitChunkSize && offset + word * kBitChunkSize < N; ++offset)
+            {
+                if ((value & (kBit1 << offset)) != 0)
+                {
+                    f(GetPoolHead() + (word * kBitChunkSize + offset));
+                }
+            }
+        }
+    }
+
+    size_t GetNumObjectsInUse() const
+    {
+        size_t count = 0;
+        ForEachActiveObject([&count](const T *) { ++count; });
+        return count;
+    }
+#endif
+
+private:
+    T * GetPoolHead() { return reinterpret_cast<T *>(mMemory); }
+    const T * GetPoolHead() const { return reinterpret_cast<const T *>(mMemory); }
+
+    using tBitChunkType                         = uintmax_t;
```

Comment:
Why this as opposed to `unsigned int`?  We should at the very least document the choice made here.  Are atomic ops on larger-than-int values just as cheap as on int-sized ones on our target hardware platforms?

Reply:
Comments added

Reply:
.. but C++11 does not guarantee that std::atomic<uintmax_t> does not require a mutex.

Even relaxed operations generate a function call with std::atomic<uintmax_t> for ARM gcc cortex-m4:
https://godbolt.org/z/T5qo5r
 
I'd recommend using "unsigned long" as this is word size on both ILP32 and LP64 platforms.

Please also assert ATOMIC_LONG_LOCK_FREE.

Reply:
Actually, why not atomic<uint8_t> ?

Reply:
> Actually, why not atomic<uint8_t> ?

Nevermind, I expect the reason is `ForEachActiveObject`.